md5_mb/aarch64/md5_mb_mgr_aarch64_sve.c

```diff
@@ -0,0 +1,210 @@
+
+  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+  OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+**********************************************************************/
+#include <stddef.h>
+#include <md5_mb.h>
+#include <assert.h>
+
+#ifndef min
+#define min(a,b)            (((a) < (b)) ? (a) : (b))
+#endif
+
+extern void md5_mb_sve(int blocks, int total_lanes, MD5_JOB **);
+extern void md5_mb_asimd_x4(MD5_JOB *, MD5_JOB *, MD5_JOB *, MD5_JOB *, int);
+extern void md5_mb_asimd_x1(MD5_JOB *, int);
+extern int md5_mb_sve_max_lanes(void);
+
+#define LANE_IS_NOT_FINISHED(state,i)  	\
+	(((state->lens[i]&(~0xff))!=0) && state->ldata[i].job_in_lane!=NULL)
+#define LANE_IS_FINISHED(state,i)  	\
+	(((state->lens[i]&(~0xff))==0) && state->ldata[i].job_in_lane!=NULL)
+#define	LANE_IS_FREE(state,i)		\
+	(((state->lens[i]&(~0xff))==0) && state->ldata[i].job_in_lane==NULL)
+#define LANE_IS_INVALID(state,i)	\
+	(((state->lens[i]&(~0xff))!=0) && state->ldata[i].job_in_lane==NULL)
+void md5_mb_mgr_init_sve(MD5_MB_JOB_MGR * state)
+{
+	unsigned int j;
+	state->unused_lanes[0] = 0x0706050403020100;
+	state->unused_lanes[1] = 0x0f0e0d0c0b0a0908;
+	state->unused_lanes[2] = 0x1716151413121110;
+	state->unused_lanes[3] = 0x1f1e1d1c1b1a1918;
+	state->num_lanes_inuse = 0;
+	for (j = 0; j < MD5_MAX_LANES; j++) {
+		state->lens[j] = j;
+		state->ldata[j].job_in_lane = 0;
+	}
+}
+
+static int md5_mb_mgr_do_jobs(MD5_MB_JOB_MGR * state)
+{
+	int lane_idx, len, i, lanes, blocks;
+	MD5_JOB *job_vecs[MD5_MAX_LANES];
+	int maxjobs;
+
+	if (state->num_lanes_inuse == 0) {
+		return -1;
+	}
+
+	maxjobs = md5_mb_sve_max_lanes();
+	if (maxjobs > MD5_MAX_LANES)
+		maxjobs = MD5_MAX_LANES;
+
+	lanes = 0;
+	len = 0;
+	for (i = 0; i < MD5_MAX_LANES && lanes < state->num_lanes_inuse; i++) {
+		if (LANE_IS_NOT_FINISHED(state, i)) {
+			if (lanes)
+				len = min(len, state->lens[i]);
+			else
+				len = state->lens[i];
+			job_vecs[lanes++] = state->ldata[i].job_in_lane;
+		}
+	}
+
+	if (lanes == 0)
+		return -1;
+	lane_idx = len & 0xff;
+	len &= ~0xff;
+	blocks = len >> 8;
+
+	// SVE is using double pipeline in parallel
+	// the least amount of jobs is single pipeline capacity minus 2
+	if (lanes >= maxjobs / 2 - 2) {
+		md5_mb_sve(blocks, lanes, job_vecs);
```

Comment:
Ok, so now I understand the `(lanes >= maxjobs / 2 - 2)` is based on real-life test on SVE-256b hardware. Would you add this information into the code comments? So far, the existing comment `// SVE is using double pipeline in parallel
	// the least amount of jobs is single pipeline capacity minus 2` explains only part of the story.

I have access to a SVE-512b machine, and I can test on it sometime next week. Whatever the result comes, I guess it only means one implementation of the Arm SVE-capable hardware. In a different micro-architecture implementation, this criteria (maxjobs/2-2) may be different. Anyway, at this moment, my best suggestion is to explain explicitly in the code comment, and people know this can be a potential tuning point.

Reply:
256-bit SVE is probably the most important sve hardware we consider at the moment. it corresponds to V1 micro-architecture used in our data center chip. we will optimize the code based on N2 (a 128-bit sve2) when we have hardware.

Reply:
Right, that's exactly what my point is. As a common _aarch64_sve.c file, at least please leave some optimization clues for software engineers from other Arm partners.